tests/test_database.py

```diff
@@ -0,0 +1,488 @@
+    "register", [abstractions.RegisterEndpoints, abstractions.register_endpoints]
+)
+def test_registration(session, models_with_kwargs, register):
+    models = models_with_kwargs[0]
+    new_app = FastAPI()
+
+    # assert that this application has no registered routes
+    assert len(registered_routes(new_app)) == 0
+
+    def get_session():
+        yield session
+
+    # register routes for this application
+    register(api=new_app, get_session=get_session, models=models)
+
+    # assert that this application now has five registered routes
+    routes = registered_routes(new_app)
+
+    assert len(routes) == 5
+
+    expected_names = ("_create", "_read_range", "_read_single", "_update", "_delete")
+    # Check names
+    for r in routes:
+        assert r.name in expected_names
+    # Check routes
+    for r in routes:
+        if r.name in ("_create", "_read_range"):
+            assert r.path == models.path
+        elif r.name in ("_read_single", "_update", "_delete"):
+            assert r.path == f"{models.path}{{id}}"
+    # Check response models
+    for r in routes:
+        if r.name in ("_create", "_read_single", "_update"):
+            assert r.response_model == models.response
+        elif r.name == "_read_range":
+            assert r.response_model == List[models.response]
+        elif r.name == "_delete":
+            assert r.response_model is None
+
+
+# Test create ---------------------------------------------------------------------------
+
+
+@pytest.mark.parametrize("entrypoint", ENTRYPOINTS)
+def test_create(session, model_to_create, entrypoint, http_server):
+    models, request, blank_opts = model_to_create
+    table = models.table(**request)
+    # make sure the database is empty
+    clear_table(session, models.table)
+
+    # get response data
+    # write respones to dict to make sure `data` is in fact collected
+    # from a specific endpoint, as opposed to carried over from a previous test call
+    # (the latter seems like it shouldn't happen, but... it seemed like it might've been?)
+    r = dict()
+    if entrypoint == "db":
+        commit_to_session(session, table)
+        # Need to `get` b/c db doesn't return a response
+        model_db = session.get(models.table, table.id)
+        data = model_db.dict()
+        r.update({entrypoint: data})
+    elif entrypoint == "abstract-funcs":
+        model_db = abstractions.create(
+            session=session, table_cls=models.table, model=models.table(**request)
+        )
+        data = model_db.dict()
+        r.update({entrypoint: data})
+    elif entrypoint == "client":
+        client = Client(base_url=http_server)
+        response = client.post(models.path, json=request)
+        assert response.status_code == 200
+        data = response.json()
+        r.update({entrypoint: data})
+    elif entrypoint == "cli":
+        data = get_data_from_cli("post", http_server, models.path, request)
+        r.update({entrypoint: data})
```

Comment:
[This pattern of test classes](https://stackoverflow.com/a/62176555/3266235) could be useful here.

Maybe we have a class for each entrypoint. Something like

```python

class EntrypointTester:
    ...
    # generic methods for every entrypoint go here

class DBTester(EntrypointTester):
    def register(...):
        ...
```

Reply:
> More generally, I would like to be able to look at the cli tests and actually see what commands are being run. Remember that the tests themselves are an important form of developer documentation.

I found it makes sense to group the tests themselves according to CRUD operation, because the evaluation of the outputs is the same for each entrypoint. This way we can have a single [`evaluate_data`](https://github.com/cisaacstern/pangeo-forge-orchestrator/blob/5141edce2557550cd12ba8da07b03d6b73218f1b/tests/test_database.py#L117-L135) static method for each test class.

But as you note, we still want to be able to approach the operations as grouped by entrypoint (for our own understanding, and as documentation). That's why I've added the `entrypoints` module which has containers for each `<Entrypoint>CRUD`, e.g. [`DatabaseCRUD`](https://github.com/cisaacstern/pangeo-forge-orchestrator/blob/5141edce2557550cd12ba8da07b03d6b73218f1b/tests/entrypoints.py#L70). These methods are then [passed through](https://github.com/cisaacstern/pangeo-forge-orchestrator/blob/5141edce2557550cd12ba8da07b03d6b73218f1b/tests/conftest.py#L186) to the fixture classes.